### PR TITLE
Task 4 - filtering and searching functionality

### PR DESCRIPTION
**Pull request by mat-kud**
There are problems with grabbing products prices, because of sponsored products

### CODE REVIEW

**Review by sabinawozniak — CHANGES_REQUESTED** [review]
Test verifyTitlesContainChosenBrand failed (maybe only on my computer, because I don't see "Corsair" on the main page): Unable to locate element: {"method":"xpath","selector":"//span[text() = 'Corsair']"}
I think that if you will add action which click "See more", then this test will pass.
And I think that you can use only the second Data Provider and use only this data which you need in the first test (only "Corsair"). It's just a suggestion - your code is ok.

**Comment by shibiza** [comment]
Hello! As I can see 

 WebElement brandNameCheckbox = webDriver.findElement(By.xpath("//span[text() = '" + brandName + "']"));

you use triple times in each of your 3 tests, so you may add this variable to your class, not inside @test method

**Comment by RoboPen** [comment]
After asking driver to expand brand list, it doesnt want to iterate over pages

**Comment by shibiza** [comment]
You know, the process finished with exit code -1. No one of your @Test works...

**Review by Petr-Kv — DISMISSED** [review]

**Comment by mat-kud** [comment]
There is a problem with verifyProductsPricesAreSorted test. I found "hidden" products in DOM structure. Thats why test fails

File: src/test/java/org/example/FilteringAndSearchingTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    private boolean verifyPricesAreInChoseRange(float minPrice, float maxPrice) {
+        boolean arePricesInChosenRange = false;
+
+        int numOfPages = webDriver.findElements(By.xpath("//span[@class='s-pagination-strip']/span[4]")).isEmpty() ?
+                1 : Integer.parseInt(webDriver.findElement(By.xpath("//span[@class='s-pagination-strip']/span[4]")).getText());
+
+        for (int i = 1; i <= numOfPages; i++) {
+//checking if prices are within specified range
+            arePricesInChosenRange = webDriver.findElements(By.xpath("//span[@class=\"a-price\"]/span[@class=\"a-offscreen\"]"))
```
Comment (sabinawozniak) [inline src/test/java/org/example/FilteringAndSearchingTest.java]: Your xpath finds elements from outside of results list which gives me false negative test result.

image

**Review by sabinawozniak — CHANGES_REQUESTED** [review]

File: src/test/java/org/example/FilteringAndSearchingTest.java (hunk 1 above)
Comment (sabinawozniak) [reply src/test/java/org/example/FilteringAndSearchingTest.java]: Oh sorry, I didn't see your previous comment, that you just found what is wrong ;) I hope that you will deal with this problem.

**Review by shibiza — APPROVED** [review]

File: src/test/java/org/example/FilteringAndSearchingTest.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+    @DataProvider(name = "brandNamesAndPriceRanges")
+    public Object[][] brandNamesAndPriceRanges() {
+        return new Object[][]{
+                {"Corsair", 50, 250}
+        };
+    }
+
+    @Test(dataProvider = "brandNamesAndPriceRanges")
+    public void verifyProductsPricesAreInDefinedRange(String brandName, float minPrice, float maxPrice) {
```
Comment (sabinawozniak) [inline src/test/java/org/example/FilteringAndSearchingTest.java:64]: image

image

File: src/test/java/org/example/FilteringAndSearchingTest.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        WebElement submitPriceRangeBtn = webDriver.findElement(By.xpath("//input[@aria-labelledby='a-autoid-1-announce']"));
+        submitPriceRangeBtn.click();
+
+        boolean arePricesInChosenRange = verifyPricesAreInChoseRange(minPrice, maxPrice);
+
+        Assert.assertTrue(arePricesInChosenRange, "Prices are not within specified range");
+    }
+
+    @Test(dataProvider = "brandNames")
+    public void verifyProductsPricesAreSortedAscendingly(String brandName) {
```
Comment (sabinawozniak) [inline src/test/java/org/example/FilteringAndSearchingTest.java:83]: Same as above test

**Review by sabinawozniak — APPROVED** [review]

**Comment by mat-kud** [comment]
Tests fail as expected, because conditions are not met. Try to comment after method and check it out by yourself

**Comment by Petr-Kv** [comment]
image

Yes, you are right

**Review by Petr-Kv — APPROVED** [review]

File: src/test/java/org/example/FilteringAndSearchingTest.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        String categoryPath = "/s?k=gaming+keyboard&pd_rd_r=da8afc49-fa94-41c3-9d45-7e811ac33b10&pd_rd_w=gSHhP&pd_rd_wg"
+                + "=fx882&pf_rd_p=12129333-2117-4490-9c17-6d31baf0582a&pf_rd_r=XYWA244WM0H05HEYD0RE&ref=pd_gw_unk";
+        webDriver.get("https://www.amazon.com" + categoryPath);
+
+        WebElement brandsExpander = webDriver.findElement(By.xpath("(//a[contains(@class,\"a-expander-header\")])[2]"));
+        JavascriptExecutor js = (JavascriptExecutor) webDriver;
+        js.executeScript("arguments[0].click();", brandsExpander);
+    }
+
+    /*@AfterTest
```
Comment (juliaHrabovska) [inline src/test/java/org/example/FilteringAndSearchingTest.java:35]: Commented code shouldn't be in the code you push to a remote repo

File: src/test/java/org/example/FilteringAndSearchingTest.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    private boolean verifyEveryTitleContainsBrandName(String brandName) {
+        WebElement paginationNextBtn;
+        boolean everyTitleContainsInputWord;
+
+        while (true) {
+            paginationNextBtn = new WebDriverWait(webDriver, Duration.ofSeconds(6))
+                    .until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[contains(@class,'s-pagination-next')]")));
+
+//checking if every title of the page contains brand name
+            everyTitleContainsInputWord = webDriver.findElements(By.xpath("//div[contains(@class,'s-card-container')]//span[contains(@class,'a-size-medium')]"))
```
Comment (juliaHrabovska) [inline src/test/java/org/example/FilteringAndSearchingTest.java:115]: It is better to make this line shorter. Just move some parts to the next line. The same for other

File: src/test/java/org/example/FilteringAndSearchingTest.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    private boolean verifyPricesInAscendingOrder(List<Float> prices) {
+        for (int i = 0; i < prices.size() - 2; i++) {
+            if (prices.get(i) > prices.get(i + 1)) {
+                return false;
+            }
+        }
+        return true;
+    }
+}
+
```
Comment (juliaHrabovska) [inline src/test/java/org/example/FilteringAndSearchingTest.java:205]: Extra lines

**Pull request merged**